### PR TITLE
feat(zod): add generateReusableSchemas opt-in flag

### LINKED ISSUE #2535: Zod Generator Produces Large Monolithic Schemas
I've observed that when using the Zod generator with an OpenAPI specification that utilises the $ref keyword to link between component schemas (e.g., a User schema referencing an Address schema under #/components/schemas), Orval recursively resolves these references.

The final Zod response schema for an operation (e.g., getUserByIdResponse) is generated as a single, large Zod object that contains the entire, inline definition of every referenced component schema, rather than simply referencing their names.

This generates massive files that are difficult to work with, especially in complex specs.

### LINKED ISSUE #3068: Zod: Apply schema suffix only to schema and not to the type
### Description

It would be nice if we could only apply the schema suffix to the schema itself and not the inferred type as well. 

### Current behavior

Currently, the suffix gets added to both schema and type.

### Proposed behavior

Only add it to the schema. 

### Proposed solution

Maybe add an option to define the type suffix?

### PR DESCRIPTION
**Pull request by z4o4z**
## Summary

Fix #3068

Adds an opt-in `override.zod.generateReusableSchemas` flag. When enabled, the Zod generator emits one reusable schema per `#/components/schemas/*` `$ref` (with `namingConvention` applied to the name) and references it everywhere instead of inlining. Cycles get `z.lazy(() => Name)` only at the edges that close them.

Default is `false`; existing output is unchanged when the flag is unset.

## Behavior

- `$ref` to `#/components/schemas/X` → an exported `const X = zod...` (or `Pet`, `pet_owner`, etc., per `namingConvention`); cross-schema references use bare identifiers, cycle-closing edges become `zod.lazy(() => X)`.
- Operation files (verb-level) import and reuse the named exports instead of inlining the schema.
- Per-operation wrapper files (`PetCreateBody.zod.ts`) are skipped when the body/response is a pure `$ref` — consumers import the component schema directly.
- Sibling fields on `$ref`:
  - **Chainable** (`nullable`, `default`, `description`) → `Pet.nullable().describe(...)` chain.
  - **Parent-driven** `optional` → `.optional()` when the parent doesn't list the property in `required`.
  - **Non-chainable** (`properties`, `example`, etc.) → falls back to inlining at that one site, with a verbose log.
- When the schemas dir is configured as a plain string (`schemas: 'src/gen/models'`) AND `client: 'zod'` AND the flag is on, the schemas dir is treated as the zod output (Zod schemas land there, not TypeScript types). Use the explicit `{ path, type: 'zod' }` form, or rely on this auto-promotion.

## Implementation

- `packages/core/src/types.ts`: adds `generateReusableSchemas?: boolean` to `ZodOptions` and required field to `NormalizedZodOptions`.
- `packages/orval/src/utils/options.ts`: defaults the flag to `false` in both normalization sites.
- `packages/zod/src/index.ts`:
  - `generateZodValidationSchemaDefinition` widened to accept `OpenApiReferenceObject`. New top-of-function branch emits a `namedRef` placeholder + chained modifiers when the input is a plain component-schema `$ref`. Non-chainable siblings fall back to today's `dereference` path at that one site.
  - The flag is threaded through every recursive call (properties, items, allOf/oneOf/anyOf, additionalProperties, tuple prefixItems/rest, multi-type).
  - `parseZodValidationSchemaDefinition` returns `usedRefs: Set<string>` and renders refs as `__REF_<name>__` sentinels.
  - `generateZodRoute` collects `usedRefs` from each parser call and rewrites the operation-level output's sentinels to bare names (operations are top-of-graph and never need lazy). `generateZod` returns the refs as `imports` with `values: true` so mode writers emit proper `import { ... } from '...'` lines.
- `packages/orval/src/reusable-schemas.ts` (new): orchestrator that walks the spec, runs the generator per root, builds the ref dependency graph, runs Tarjan's SCC + DFS back-edge detection, rewrites sentinels into direct refs or `zod.lazy(...)`, and emits in topological order.
- `packages/orval/src/write-zod-specs.ts`: `writeZodSchemas` and `generateZodSchemasInline` route through the orchestrator when the flag is on. `writeZodSchemasFromVerbs` skips pure-ref wrappers.
- `packages/orval/src/write-specs.ts`: when `client: 'zod'` + flag is on + `schemas:` is a plain string, treat the schemas dir as zod (auto-promotion to the zod writer).

## Test plan

- [x] Unit tests for `generateZodValidationSchemaDefinition` cover plain refs, all chainable siblings, parent-driven optional, and the non-chainable fall-back.
- [x] Unit tests for `parseZodValidationSchemaDefinition` cover `__REF__` sentinel emission and `usedRefs` reporting.
- [x] Unit tests for the orchestrator (`reusable-schemas.test.ts`): name resolution + conflict guard, ref collection, per-root generation with sentinels, Tarjan SCC + back-edge cases (DAG, simple cycle, self-loop, multi-node SCC), and sentinel rewriting + topological order.
- [x] Integration tests in `write-zod-specs.test.ts` for cross-file imports (`schemas: { type: 'zod' }`), wrapper-skip for pure refs, and inline orchestrator output.
- [x] End-to-end snapshot via `samples/swr-with-zod`'s new `petstoreZodReusable` config (uses `-reusable.zod.ts` file extension so its output sits alongside the existing TS petstore + plain zod outputs without colliding).
- [x] All existing tests pass; default-off keeps existing output unchanged.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a Zod generation option generateReusableSchemas (disabled by default). When enabled, component schemas are emitted as reusable named exports and referenced across generated files; cycles use lazy wrappers.

* **Tests**
  * Added suites covering naming, reachability, dependency ordering, lazy-edge handling, rewriting, and end-to-end reusable-schema emission.

* **Documentation**
  * Documented generateReusableSchemas and updated examples.

* **Samples**
  * Added sample config and generated reusable Zod schema snapshots.

* **Chores**
  * Default schema file extension normalized to .ts in outputs/tests.

[Review Change Stack](https://app.coderabbit.ai/change-stack/orval-labs/orval/pull/3449?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by melloware — CHANGES_REQUESTED** [review]
@z4o4z make sure to update the `/docs` for Zod for this new option

**Comment by melloware** [comment]
i believe this fixes https://github.com/orval-labs/orval/issues/2535

File: packages/zod/src/index.ts [hunk 1]
```diff
@@ -1520,8 +1647,30 @@ export const parseParameters = ({
...
+        ? (() => {
+            const raw = resolveRef<OpenApiSchemaObject>(
+              parameter.schema,
+              context,
+            ).schema;
+            // Merge in the parameter description without mutating the shared ref.
+            if (parameter.description) {
+              return Object.assign({}, raw, {
+                description: parameter.description,
+              }) as OpenApiSchemaObject;
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/zod/src/index.ts]: Valid — fixed in 670a255. `parseParameters` now preserves `$ref` parameter schemas verbatim (with `description` shallow-merged) so the namedRef path can fire. Removed the `resolveRef` call that was stripping the reference.

File: packages/orval/src/write-zod-specs.ts [hunk 2]
```diff
@@ -415,7 +566,9 @@ export async function writeZodSchemasFromVerbs(
         ? [
             {
               name: `${pascal(verbOption.operationName)}Body`,
-              schema: dereference(bodySchema, zodContext),
+              schema: useReusableSchemas
+                ? bodySchema
+                : dereference(bodySchema, zodContext),
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts:571]: Valid — fixed in e3c68ae. `writeZodSchemasFromVerbs` now passes `useReusableSchemas` (and the equivalent param to `generateFormDataZodSchema`) into the inner generator calls so nested refs inside non-pure-ref body/response wrappers emit namedRef placeholders correctly.

File: packages/orval/src/write-zod-specs.ts [hunk 3]
```diff
@@ -440,7 +593,12 @@ export async function writeZodSchemasFromVerbs(
                     .filter((p) => 'schema' in p && p.schema)
                     .map((p) => [
                       p.name,
-                      dereference(p.schema as OpenApiSchemaObject, zodContext),
+                      useReusableSchemas
+                        ? (p.schema as OpenApiSchemaObject)
+                        : dereference(
+                            p.schema as OpenApiSchemaObject,
+                            zodContext,
+                          ),
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts:601]: Valid — fixed in e3c68ae. `writeZodSchemasFromVerbs` now passes `useReusableSchemas` (and the equivalent param to `generateFormDataZodSchema`) into the inner generator calls so nested refs inside non-pure-ref body/response wrappers emit namedRef placeholders correctly.

File: packages/orval/src/write-zod-specs.ts [hunk 4]
```diff
@@ -468,7 +626,12 @@ export async function writeZodSchemasFromVerbs(
...
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts:634]: Valid — fixed in e3c68ae. `writeZodSchemasFromVerbs` now passes `useReusableSchemas` (and the equivalent param to `generateFormDataZodSchema`) into the inner generator calls so nested refs inside non-pure-ref body/response wrappers emit namedRef placeholders correctly.

File: packages/orval/src/write-zod-specs.ts [hunk 5]
```diff
@@ -498,7 +661,9 @@ export async function writeZodSchemasFromVerbs(
           )
           .map((responseType) => ({
             name: responseType.value,
-            schema: dereference(responseType.originalSchema, zodContext),
+            schema: useReusableSchemas
+              ? responseType.originalSchema
+              : dereference(responseType.originalSchema, zodContext),
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts:666]: Valid — fixed in e3c68ae. `writeZodSchemasFromVerbs` now passes `useReusableSchemas` (and the equivalent param to `generateFormDataZodSchema`) into the inner generator calls so nested refs inside non-pure-ref body/response wrappers emit namedRef placeholders correctly.

File: packages/orval/src/write-specs.ts [hunk 6]
```diff
@@ -277,9 +277,40 @@ export async function writeSpecs(
...
+    const schemasPath = isString(output.schemas)
+      ? output.schemas
+      : output.schemas.path;
+    const isZodSchemas =
+      (!isString(output.schemas) && output.schemas.type === 'zod') ||
+      (output.client === 'zod' &&
+        output.override.zod.generateReusableSchemas === true);
+
+    if (isZodSchemas) {
+      const fileExtension = output.fileExtension || '.zod.ts';
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-specs.ts]: Valid — fixed in e3c68ae. The normalizer always sets `output.fileExtension` (default `.ts`), so the previous `|| '.zod.ts'` fallback was dead. New logic treats the bare default as unset and emits `.zod.ts`; any explicitly user-set extension is honoured.

File: packages/orval/src/reusable-schemas.ts [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+};
+
+/**
+ * Convert a single `#/components/schemas/X` ref into the export name we will
+ * emit for it: the last `$ref` segment with `namingConvention` applied.
+ */
+export const resolveSchemaName = (
+  ref: string,
+  namingConvention: NamingConvention,
+): string => conventionName(lastRefSegment(ref), namingConvention);
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/reusable-schemas.ts:26]: Valid — fixed in e3c68ae. `resolveSchemaNames` now validates each converted name against the JS identifier regex `^[A-Za-z_$][A-Za-z0-9_$]*$` and throws a clear error for `kebab-case` (and any other convention that produces dashes). Docs updated to call out `camelCase`/`PascalCase`/`snake_case` as the supported conventions when this flag is enabled. Existing kebab-case test replaced with a snake_case round-trip + a new test asserting the throw.

File: packages/zod/src/index.ts [hunk 8]
```diff
@@ -1865,18 +2052,25 @@ export const ${operationResponse} = zod.array(${operationResponse}Item)${
...
   return {
     implementation: implementation ? `${implementation}\n\n` : '',
-    imports: [],
+    // Zod schemas are runtime values (not type-only), so mark with values: true
+    // to prevent the import writer from emitting `import type { ... }`.
+    imports: [...(usedRefs ?? [])].map((name) => ({
+      name,
+      schemaName: name,
+      values: true,
+    })),
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/zod/src/index.ts]: Valid — fixed in e3c68ae. `[...usedRefs].sort()` is now applied before mapping into the imports array so emission order is deterministic across runs regardless of Set insertion order.

File: packages/orval/src/reusable-schemas.ts [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+  }
+  const record = value as Record<string, unknown>;
+  if (isComponentSchemaRef(record.$ref)) {
+    refs.add(record.$ref);
+    // Do not descend through a $ref; the BFS will follow it via components/schemas.
+    return;
+  }
+  for (const key of Object.keys(record)) {
+    collectRefsInValue(record[key], refs);
+  }
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/reusable-schemas.ts:98]: Valid — fixed in 670a255. `collectRefsInValue` no longer returns early after recording a `$ref`; it skips descending into the `$ref` string itself but continues iterating sibling keys so OpenAPI 3.1 ref siblings are walked correctly.

File: packages/orval/src/reusable-schemas.ts [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+    };
+
+    const start = [...scc].sort()[0];
+    dfs(start);
+
+    // Some nodes may not be reachable from `start` in the digraph; visit them too.
+    for (const node of scc) {
+      if (!visited.has(node)) dfs(node);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/reusable-schemas.ts]: Valid (and important — thanks for catching this). Fixed in 670a255 by integrating back-edge detection into Tarjan itself: the SCC nodes are returned in reverse DFS-finish order (Tarjan's natural pop order), and lazy edges are recorded as Tarjan visits them. Both the lazy classification and the emit order come from a single DFS, so non-lazy edges always point to already-emitted nodes — no more TDZ risk. The lex-sort in `topoSortEntries` is removed.

File: packages/orval/src/write-specs.ts [hunk 11]
```diff
@@ -277,9 +277,40 @@ export async function writeSpecs(
...
 
   if (output.schemas) {
-    if (isString(output.schemas)) {
+    const schemasPath = isString(output.schemas)
+      ? output.schemas
+      : output.schemas.path;
+    const isZodSchemas =
+      (!isString(output.schemas) && output.schemas.type === 'zod') ||
+      (output.client === 'zod' &&
+        output.override.zod.generateReusableSchemas === true);
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-specs.ts]: Valid — fixed in 670a255. The auto-promote-to-zod branch is now gated on `isString(output.schemas)` so an explicit `{ type: 'typescript' }` config is kept on the TS path even when `client: 'zod'` + the flag are set. The bare-object form must opt in via `{ type: 'zod' }`.

File: packages/orval/src/write-zod-specs.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts:571]: Already addressed by the threading fix in e3c68ae. `writeZodSchemasFromVerbs` now passes `useReusableSchemas` (and the form-data equivalent) into the inner generator calls so nested refs inside non-pure-ref wrappers do emit namedRef placeholders and resolve to the shared exports.

File: packages/zod/src/index.ts [hunk 12]
```diff
@@ -1404,10 +1521,17 @@ const parseBodyAndResponse = ({
...
+  // directly to generateZodValidationSchemaDefinition so that component schema
+  // references are emitted as named identifiers instead of being inlined.
+  const effectiveSchema: OpenApiSchemaObject | OpenApiReferenceObject =
+    useReusableSchemas
+      ? parseType === 'body'
+        ? removeReadOnlyProperties(schema as OpenApiSchemaObject)
+        : schema
+      : parseType === 'body'
+        ? removeReadOnlyProperties(resolvedJsonSchema)
+        : resolvedJsonSchema;
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/zod/src/index.ts:1545]: Valid as a behavioral observation, but it's an intentional trade-off of opt-in reuse rather than a bug to fix at the call site. When you opt into shared schemas, request and response endpoints reference the same exported component schema, so `readOnly` properties remain in the body validator (without the flag, `dereference + removeReadOnlyProperties` strips them at body-generation time). Stripping the properties would either inline (defeating reuse) or fork the export into request-side and response-side variants (doubling the surface). Documented as a caveat in `docs/content/docs/reference/configuration/output.mdx` (commit 670a255) so users know to avoid `readOnly` on schemas they share between requests and responses, or split the schema in the OpenAPI source.

File: packages/zod/src/index.ts [hunk 13]
```diff
@@ -1520,8 +1647,30 @@ export const parseParameters = ({
...
+              context,
+            ).schema;
+            // Merge in the parameter description without mutating the shared ref.
+            if (parameter.description) {
+              return Object.assign({}, raw, {
+                description: parameter.description,
+              }) as OpenApiSchemaObject;
+            }
+            return raw;
+          })()
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/zod/src/index.ts]: Already addressed in e3c68ae. `parseParameters` now preserves the `$ref` (shallow-merging `description` without mutating the shared ref object) when `useReusableSchemas` is on, so parameter refs reach the namedRef path and emit as reusable exports.

File: packages/orval/src/write-specs.ts [hunk 14]
```diff
@@ -277,9 +277,44 @@ export async function writeSpecs(
...
+
+    if (isZodSchemas) {
+      // The normalizer always sets `output.fileExtension` (default `.ts`), so
+      // we can't use `||` to detect "unset". Treat the bare normalizer default
+      // as unset and fall back to `.zod.ts`; honour any user-set extension.
+      const fileExtension =
+        output.fileExtension === '.ts' ? '.zod.ts' : output.fileExtension;
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-specs.ts]: Valid — fixed properly in 5ed3dff. Moved the default-extension logic into the normalizer where we still know whether the user explicitly set `fileExtension`. The check is now: if the output is zod-schemas (either explicit `{ type: 'zod' }` or string `schemas:` + `client: 'zod'` + `generateReusableSchemas`), default `.zod.ts`; otherwise `.ts`. A user-set extension still wins via the existing `outputOptions.fileExtension ?? defaultFileExtension` chain. The workaround in `write-specs.ts` is removed.

**Comment by z4o4z** [comment]
> i believe this fixes #2535

yep

File: packages/orval/src/utils/options.ts [hunk 15]
```diff
@@ -218,7 +218,18 @@ export async function normalizeOptions(
...
+  // only distinguish "user explicitly set `.ts`" from "user said nothing"
+  // here — once normalization runs we'd be guessing.
+  const isZodSchemasOutput =
+    !!outputOptions.schemas &&
+    ((!isString(outputOptions.schemas) &&
+      outputOptions.schemas.type === 'zod') ||
+      (isString(outputOptions.schemas) &&
+        (outputOptions.client ?? client) === 'zod' &&
+        outputOptions.override?.zod?.generateReusableSchemas === true));
+  const defaultFileExtension = isZodSchemasOutput ? '.zod.ts' : '.ts';
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/utils/options.ts]: Valid — fixed in cd9cbdb5e1ba0c7d4ed34d873cc61b85b2f54678. Introduced a separate `output.schemaFileExtension` field that defaults to `.zod.ts` for zod-schemas outputs and `.ts` otherwise. The global `output.fileExtension` keeps its `.ts` default so non-schema writers (mode writers, mock writers, the workspace barrel) aren't dragged into the zod world. A user-set `fileExtension` still flows to both.

File: packages/zod/src/index.ts [hunk 16]
```diff
@@ -209,10 +216,88 @@ export const generateZodValidationSchemaDefinition = (
...
+    const allSiblingsChainable = siblings.every(isChainable);
+
+    if (allSiblingsChainable) {
+      const refName = conventionName(
+        getRefInfo(schema.$ref, context).originalName,
+        context.output.namingConvention,
+      );
+      const functions: [string, unknown][] = [
+        ['namedRef', { name: refName, sourceRef: schema.$ref }],
+      ];
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/zod/src/index.ts:249]: Valid — fixed in cd9cbdb5e1ba0c7d4ed34d873cc61b85b2f54678. The orchestrator's `lastRefSegment` now mirrors `getRefInfo(...).originalName`: it URL-decodes the segment and unescapes the RFC 6901 tokens `~1` → `/` and `~0` → `~`. Both the namedRef emission and the orchestrator's export-name resolution arrive at identical names for refs with escaped characters.

File: packages/zod/src/index.ts [hunk 17]
```diff
@@ -209,10 +216,88 @@ export const generateZodValidationSchemaDefinition = (
...
+
+      // .default(...) — emit a const for non-primitive defaults to match the rest of the generator.
+      if (refSchema.default !== undefined) {
+        const defaultVarName = `${name}Default`;
+        const defaultLiteral = stringify(refSchema.default);
+        if (defaultLiteral !== undefined) {
+          consts.push(`export const ${defaultVarName} = ${defaultLiteral};`);
+          functions.push(['default', defaultVarName]);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/zod/src/index.ts]: Valid — fixed in cd9cbdb5e1ba0c7d4ed34d873cc61b85b2f54678. The ref-default branch now consults `rules.constNameRegistry` and applies the same word-counter suffix used elsewhere in `generateZodValidationSchemaDefinition`, so multiple defaults sharing `name` produce `<name>Default`, `<name>DefaultOne`, `<name>DefaultTwo`, ... without colliding.

File: packages/orval/src/reusable-schemas.test.ts [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+        ['b', new Set(['a', 'c'])],
+        ['c', new Set(['a'])],
+      ]),
+    );
+    expect(edges.size).toBeGreaterThan(0);
+    expect(edges.size).toBeLessThan(5);
+  });
+});
+
+import { rewriteReusableSchemas } from './reusable-schemas';
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/reusable-schemas.test.ts]: Valid — fixed in cd9cbdb5e1ba0c7d4ed34d873cc61b85b2f54678. `rewriteReusableSchemas` is now imported with the other named exports from `./reusable-schemas` at the top of the test file. The mid-file import is gone.

File: packages/orval/src/reusable-schemas.ts [hunk 19]
```diff
@@ -0,0 +1,349 @@
...
+    reachable.add(ref);
+    queue.push(ref);
+  }
+
+  const componentSchemas = (spec.components?.schemas ?? {}) as Record<
+    string,
+    unknown
+  >;
+  while (queue.length > 0) {
+    const currentRef = queue.shift() as string;
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/reusable-schemas.ts]: Valid — fixed in cd9cbdb5e1ba0c7d4ed34d873cc61b85b2f54678. The BFS now uses an index-based loop (`for (let i = 0; i < queue.length; i++)`) instead of `queue.shift()`, dropping the traversal from O(n²) to O(n).

**Review by melloware — CHANGES_REQUESTED** [review]
looks like a lint issue

File: packages/orval/src/utils/options.ts [hunk 20]
```diff
@@ -291,6 +309,8 @@ export async function normalizeOptions(
       namingConvention:
         outputOptions.namingConvention ?? NamingConvention.CAMEL_CASE,
       fileExtension: outputOptions.fileExtension ?? defaultFileExtension,
+      schemaFileExtension:
+        outputOptions.fileExtension ?? defaultSchemaFileExtension,
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/utils/options.ts]: Valid — fixed in e3292ae9fa722e6cc058d57ce461134066a8978e. Added `schemaFileExtension?: string` to `OutputOptions` and changed the normalizer to honor it: `outputOptions.schemaFileExtension ?? outputOptions.fileExtension ?? defaultSchemaFileExtension`. Users can now set a schema-specific extension independently of the global one.

File: packages/zod/src/index.ts [hunk 21]
```diff
@@ -209,10 +216,94 @@ export const generateZodValidationSchemaDefinition = (
...
+
+      // .describe('...')
+      if (typeof refSchema.description === 'string') {
+        functions.push(['describe', `"${escape(refSchema.description)}"`]);
+      }
+
+      // .optional() — driven by the parent's `required`, not a $ref sibling.
+      if (!(rules?.required ?? false)) {
+        functions.push(['optional', undefined]);
+      }
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/zod/src/index.ts]: Valid — fixed in e3292ae9fa722e6cc058d57ce461134066a8978e. Reordered the ref-branch to match the main path: nullable/nullish/optional first, then default, then describe. `.optional()` is now skipped when a default is set (matching the existing non-reusable behaviour at line ~932). Tests added covering the default+optional skip and the nullable+!required → `.nullish()` combination.

File: packages/orval/src/write-zod-specs.ts [hunk 22]
```diff
@@ -276,13 +289,72 @@ export function generateZodSchemasInline(
...
+      return (
+        `${consts}export const ${entry.name} = ${entry.zod}\n\n` +
+        `export type ${entry.name} = zod.input<typeof ${entry.name}>;\n` +
+        `export type ${entry.name}Output = zod.output<typeof ${entry.name}>;`
+      );
+    })
+    .join('\n\n');
+
+  return `import { z as zod } from 'zod';\n\n${body}\n`;
+}
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts:336]: Valid — fixed in e3292ae9fa722e6cc058d57ce461134066a8978e. `generateReusableSchemaSet` now expands the seed list to the transitive closure: it builds a name→ref map across all component schemas, then iteratively scans each generated entry's `usedRefs` for names that aren't yet generated and adds the corresponding refs to the work queue. Loop runs until stable, so refs that are reachable from any root but not directly listed are still emitted. Test added: seed with just `Pet`, expect both `Pet` and the transitively-referenced `Owner` in the result.

File: packages/orval/src/write-zod-specs.ts [hunk 23]
```diff
@@ -276,13 +289,72 @@ export function generateZodSchemasInline(
...
+    coerce,
+  });
+
+  const rewritten = rewriteReusableSchemas(entries);
+
+  const body = rewritten
+    .map((entry) => {
+      const consts = entry.consts ? `${entry.consts}\n\n` : '';
+      return (
+        `${consts}export const ${entry.name} = ${entry.zod}\n\n` +
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts]: Valid — fixed in 6e1bd6da197e92c1e47beb310e4e95d2e8a39531. Added an explicit `;` after `${entry.zod}` in this inline writer. Note: prettier-formatted output already had the trailing semicolon (it normalizes the end of the chain), but emitting it in the raw string keeps consumers without a formatter consistent and avoids ASI edge cases.

File: packages/orval/src/write-zod-specs.ts [hunk 24]
```diff
@@ -358,6 +430,83 @@ export async function writeZodSchemas(
...
+        const importedFile = conventionName(r, output.namingConvention);
+        return `import { ${r} } from './${importedFile}${importExt}';`;
+      })
+      .join('\n');
+
+    const consts = entry.consts ? `${entry.consts}\n\n` : '';
+    const fileContent =
+      `${header}import { z as zod } from 'zod';\n` +
+      (imports ? `${imports}\n\n` : '\n') +
+      `${consts}export const ${entry.name} = ${entry.zod}\n\n` +
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts]: Valid — fixed in 6e1bd6da197e92c1e47beb310e4e95d2e8a39531. Same change applied to the per-file reusable writer.

File: docs/content/docs/reference/configuration/output.mdx [hunk 25]
```diff
@@ -1050,6 +1051,22 @@ Configure Zod datetime/time validation options. `dateTimeOptions` defaults to `{
...
+
+**Type:** `boolean`
+
+**Default:** `false`
+
+Emit one reusable Zod schema per OpenAPI `#/components/schemas/*` `$ref` instead of inlining. The exported name is the last `$ref` segment with `namingConvention` applied. Other schemas and operation files reference the export by name (cycles are wrapped in `zod.lazy(() => Name)` only on the edges that close them).
+
+Behavior:
+
+- When `schemas:` is configured (string or `{ type: 'zod' }`) and `client: 'zod'` is set, the schemas directory holds reusable Zod schemas instead of TypeScript types. Set `fileExtension` to disambiguate output if you keep both TS types and reusable Zod schemas in the same directory.
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply docs/content/docs/reference/configuration/output.mdx]: Valid — fixed in 6e1bd6da197e92c1e47beb310e4e95d2e8a39531. Added a top-level `## schemaFileExtension` section to the output docs explaining the new option, with an example showing how to keep client files at `.ts` while schema files are emitted as `.zod.ts`. The reusable-schemas bullet now links to it instead of pointing at `fileExtension`.

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
This also fixed #3068
